src/hal/src/device.rs

```diff
@@ -714,6 +714,56 @@ pub trait Device<B: Backend>: fmt::Debug + Any + Send + Sync {
     /// validation layers that can print a friendly name when referring to objects in error messages
     unsafe fn set_pipeline_layout_name(&self, pipeline_layout: &mut B::PipelineLayout, name: &str);
 
+    /// Create external buffer
+    unsafe fn create_external_buffer(
+        &self,
+        external_memory_type_flags: external_memory::ExternalMemoryTypeFlags,
+        usage: buffer::Usage,
+        sparse: memory::SparseFlags,
+        size: u64,
+    ) -> Result<B::Buffer, external_memory::ExternalBufferCreateError>;
+
+    /// Allocate external memory
+    unsafe fn allocate_exportable_memory(
+        &self,
+        external_memory_types: external_memory::ExternalMemoryTypeFlags,
+        dedicated_allocation: Option<external_memory::BufferOrImage<B>>,
+        mem_type: MemoryTypeId,
+        size: u64,
+    ) -> Result<B::Memory, external_memory::ExternalMemoryAllocateError>;
+
+    /// Import external memory
+    unsafe fn import_external_memory(
+        &self,
+        external_memory: external_memory::ExternalMemory,
+        dedicated_allocation: Option<external_memory::BufferOrImage<B>>,
+        mem_type: MemoryTypeId,
+    ) -> Result<B::Memory, external_memory::ExternalMemoryAllocateError>;
+
+    #[cfg(any(unix))]
+    #[cfg_attr(feature = "unstable", doc(cfg(windows)))]
+    /// Export memory as unix file descriptor
+    unsafe fn export_memory_as_fd(
+        &self,
+        external_memory_type: external_memory::ExternalMemoryFdType,
+        memory: &B::Memory,
+    ) -> Result<std::os::unix::io::RawFd, external_memory::ExternalMemoryExportError>;
+
+    #[cfg(any(windows))]
+    #[cfg_attr(feature = "unstable", doc(cfg(windows)))]
+    /// Export memory as windows handle
+    unsafe fn export_memory_as_handle(
```

Comment:
There is quite a few new entry points and types. Is all of this necessary? Can we find a way to reduce the API surface to the minimum that would be sufficient?

Reply:
With the new code the patform specific functions have been merged into the `export_memory` function. Now the whole external memory api is composed by 1 physical device function and 3 device function (2 of them are `Buffer` specific, so there will be 2 more for the `Image` counterpart). Maybe could be possible to pass both buffer or image to the same function using an enum, so that the same function could be used for both buffer and image. In this way there will be only 3 device functions.